### PR TITLE
Optimize cchq_prbac_bootstrap

### PR DESCRIPTION
**Pull request by millerdev**
Local profile: 8.5s -> 0.05s
Queries:
- before: 544 (depends on number of roles/grants)
- after: 6 (could be 5 with Django 1.10)

Looks like there are 9 migrations that call this, and it's also called by `prepare_domain` which is used extensively in ilsgateway and ewsghana tests. Also, this is called by `instantiate_accounting`, which is used in many HQ tests.

### CODE REVIEW

**Comment by millerdev** [comment]
@dimagi/test-team

**Comment by benrudolph** [comment]
Hero

File: corehq/apps/accounting/utils.py [hunk 1]
```diff
@@ -275,52 +275,55 @@ def get_default_domain_url(domain):
...
-def remove_grant(priv_slug, dry_run=False):
-    grants = Grant.objects.filter(to_role__slug=priv_slug)
-    if dry_run:
-        if grants:
-            logger.info("[DRY RUN] Removing privilege %s", priv_slug)
-    else:
-        if grants:
-            grants.delete()
-            logger.info("Removing privilege %s", priv_slug)
+def log_removed_grants(priv_slugs, dry_run=False):
```
Comment (czue) [inline corehq/apps/accounting/utils.py]: :+1: to the rename as well

**Review by czue — APPROVED** [review]
this is amazing!

I found it pretty hard to read the diff closely enough to convince myself it's doing the same thing as before, but assume it's well tested? might be good for @nickpell to give this a quick look as well.

**Review by emord — APPROVED** [review]
@NoahCarnahan during the hackathon you mentioned accounting hits sql 80k times. I wonder how much this improves that

**Review by nickpell — APPROVED** [review]
looks great!

**Comment by NoahCarnahan** [comment]
Yeah I'm excited to see the outcome of this! I believe accounting made like 80k queries total, and something like half of them were django prbac queries.

**Comment by emord** [comment]
Out of curiosity I ran the numbers for ilsgateway tests:

no changes: `Ran 122 tests in 613.970s`
only daniel's change: `Ran 122 tests in 467.417s`
only my change: `Ran 122 tests in 368.230s`
our changes together: `Ran 122 tests in 291.860s`

Looks like ilsgateway may be quite a bit lower after on our test timing graphs after today

**Comment by emord** [comment]
real failure:

```
======================================================================
FAIL: When there is no --dry-run passed, it defaults to false, and
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/hqadmin/tests/test_prbac.py", line 74, in test_non_dry_run
    self.assertGreater(Grant.objects.count(), 10)
AssertionError: 0 not greater than 10
----------------------------------------------------------------------
```

File: corehq/apps/hqadmin/management/commands/cchq_prbac_bootstrap.py [hunk 2]
```diff
@@ -53,39 +53,47 @@ def handle(self, dry_run=False, verbose=False, fresh_start=False, testing=False,
...
-        for role in self.BOOTSTRAP_PRIVILEGES + self.BOOTSTRAP_PLANS:
-            self.ensure_role(role, dry_run=dry_run)
+        self.roles_by_slug = {role.slug: role for role in Role.objects.all()}
+        self.ensure_roles(self.BOOTSTRAP_PRIVILEGES + self.BOOTSTRAP_PLANS, dry_run)
 
-        for (plan_role_slug, privs) in self.BOOTSTRAP_GRANTS.items():
-            for priv_role_slug in privs:
-                ensure_grant(plan_role_slug, priv_role_slug, dry_run=dry_run, verbose=self.verbose)
+        ensure_grants(
+            self.BOOTSTRAP_GRANTS.items(), # py3 iterable
```
Comment (dimagimon) [inline corehq/apps/hqadmin/management/commands/cchq_prbac_bootstrap.py]: E261 at least two spaces before inline comment

**Comment by czue** [comment]
holy moly!
image

**Pull request merged**